task/common/resource.go

```diff
@@ -0,0 +1,28 @@
+package common
+
+import (
+	"context"
+)
+
+// BlobContainer defines the interface implemented by blob container resources
+// across cloud providers.
+type BlobContainer interface {
+	Resource
+	// Identifier returns the identifier of the container.
+	Identifier() string
+}
+
+// Resource defines the interface implemented by deployment resources.
+type Resource interface {
+	Read(ctx context.Context) error
+	Create(ctx context.Context) error
+	Delete(ctx context.Context) error
+}
+
+// StorageCredentials is an interface implemented by data sources and resources
+// that provide access to cloud storage buckets.
+type StorageCredentials interface {
+	// ConnectionString returns the connection string necessary to access
+	// an S3 bucket.
+	ConnectionString(ctx context.Context) (string, error)
+}
```

Comment:
Can we assume that there is no need for separate credentials and the storage can be accessed with the same credentials provided to manage the rest of resources? In cases where different credentials are needed (e.g. Microsoft Azure), they can be retrieved using the main credentials:

https://github.com/iterative/terraform-provider-iterative/blob/66e18e4dcacc798b7d9a33a686fb991be4c37eda/task/az/resources/resource_storage_account.go#L70

Reply:
So, using cloud credentials to retrieve azure storage credentials will only work if the storage is owned by the same azure account. But that is not necessarily the case. I was thinking of also supporting azure sas url's which allow granting access to storage containers by sharing a single url (albeit a lengthy one).

The other side of the issue (and probably the main reason why I moved the connection string generation out to this interface) is separation of concerns. The container resource/data source is the right component to handle generation of the rclone connection string.

Reply:
> _So, using cloud credentials to retrieve azure storage credentials will only work if the storage is owned by the same azure account._

What do you mean by account? 🤔

Reply:
By `account` I meant the subscription - sorry about not being clear enough.

Reply:
> https://github.com/iterative/terraform-provider-iterative/blob/66e18e4dcacc798b7d9a33a686fb991be4c37eda/task/az/resources/resource_storage_account.go#L70
This only works if the storage account is in the same resource group. In our case (since we've just created the resource group) we'd need to pass the resource group that the already existing storage account belongs to.

Reply:
It seems to me a reasonable requirement. If the storage account were in a different subscription, it would have to be treated as an user-controlled storage (e.g. DVC remote) and not as a task storage.

I'd prefer to keep task storage as local and native as possible, and leave other kinds of data synchronization to specialized tools.

Reply:
ok, this still leaves the issue of the storage account and its access key

Reply:
> > _the issue of the storage account and its access key_
> 
> question
Well, in order to access a blob container, you need a storage account. And to retrieve a storage account's key you need to know its resource group, so either way the user needs to supply either `(storage_account, key)` or `(resource_group, storage_account)`.